streamer/pipe.py

```diff
@@ -0,0 +1,146 @@
+# Copyright 2019 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""A module that encapsulates all the platform-specific logic related to creating
+named pipes."""
+
+import os
+import uuid
+from threading import Thread
+from typing import Any
+
+class Pipe(Thread):
+  """A class that represents a pipe."""
+
+  def __init__(self):
+    """Initializes a non-functioning pipe."""
+    
+    self._read_pipe_name = ''
+    self._write_pipe_name = ''
+
+    # Windows specific declarations.
+    self.buf_size = 0
+    self._read_side: Any = None
+    self._write_side: Any = None
+
+  @staticmethod
+  def create_ipc_pipe(temp_dir: str, suffix: str = '') -> 'Pipe':
+    """A static method used to create a pipe between two processes. 
+    
+    On POXIS systems, it creates a named pipe using `os.mkfifo`.
+    
+    On Windows platforms, it starts a backgroud thread that transfars data from the
+    writer to the reader process it is connected to.
+    """
+
+    unique_name = str(uuid.uuid4()) + suffix
+    pipe = Pipe()
+
+    # New Technology, aka WindowsNT.
+    if os.name == 'nt':
+      import win32pipe # type: ignore
+      # Initialize the pipe object as a thread with `daemon` attribute set
+      # to True so that the thread shuts down when the caller thread exits.
+      Thread.__init__(pipe, daemon=True)
```

Comment:
It feels kind of hacky to me to suddenly make Pipe a Thread.  This feels like it should be more of a "has-a" relationship than "is-a".  It appears that the run() method only uses three members of "Pipe": _read_side, _write_side, and buf_size.  What about something like:

```py
self._thread = Thread(
    target=Pipe._thread_fn,
    args=(self._read_side, self._write_side, self._buf_size),
    daemon=True)
self._thread.start()
```

You could rename `run()` to `@staticmethod _thread_fn(read_side, write_side, buf_size)`.

Reply:
Yes, that makes more sense since the `Pipe` object now represents a POSIX/Win pipe.